src/journal/compress.c

```diff
@@ -668,12 +825,197 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
 #endif
 }
 
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *s;
+        size_t ret;
+        uint64_t left = max_bytes, in_bytes = 0;
+        int r;
+        ssize_t ss;
+        size_t out_size, in_size;
+        _cleanup_free_ char *out;
+        char *in;
+        ZSTD_outBuffer ZSTD_out;
+        ZSTD_inBuffer ZSTD_in;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        s = ZSTD_createCStream();
+        if (!s)
+                return -ENOMEM;
+        out_size = ZSTD_CStreamOutSize();
+        in_size = ZSTD_CStreamInSize();
+        out = malloc(out_size + in_size);
+        if (!out)
+                return -ENOMEM;
+        in = out + out_size;
+        ZSTD_in.src = in;
+        ZSTD_in.size = out_size;
+        ZSTD_in.pos = 0;
+        ZSTD_out.dst = out;
+        ZSTD_out.size = out_size;
+        ZSTD_out.pos = 0;
+
+        ret = ZSTD_initCStream(s, ZSTD_COMPRESSION_LEVEL);
+        if (ZSTD_isError(ret) > 0) {
+                log_debug("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errno(ret);
+        }
+
+        for (;;) {
+                ss = read(fdf, in, in_size);
```

Comment:
Might be only me, but I would define variables that arent needed outside within the scope
```
ZSTD_inBuffer ZSTD_in = { .src = in, .pos = 0 };
```